rust/index/src/spann/types.rs

```diff
@@ -210,11 +269,350 @@ impl SpannIndexWriter {
             }
             None => Self::create_posting_list(blockfile_provider).await?,
         };
+
+        let max_head_id = match max_head_id_bf_id {
+            Some(max_head_id_bf_id) => {
+                let reader = blockfile_provider
+                    .read::<&str, u32>(max_head_id_bf_id)
+                    .await;
+                match reader {
+                    Ok(reader) => reader
+                        .get("", MAX_HEAD_OFFSET_ID)
+                        .await
+                        .map_err(|_| {
+                            SpannIndexWriterConstructionError::BlockfileReaderConstructionError
+                        })?
+                        .unwrap(),
+                    Err(_) => 1,
+                }
+            }
+            None => 1,
+        };
         Ok(Self::new(
             hnsw_index,
             hnsw_provider.clone(),
+            blockfile_provider.clone(),
             posting_list_writer,
+            max_head_id,
             versions_map,
+            distance_function,
+            dimensionality,
         ))
     }
+
+    fn add_versions_map(&self, id: u32) -> u32 {
+        // 0 means deleted. Version counting starts from 1.
+        let mut write_lock = self.versions_map.write();
+        write_lock.versions_map.insert(id, 1);
+        *write_lock.versions_map.get(&id).unwrap()
+    }
+
+    #[allow(dead_code)]
+    async fn rng_query(
```

Comment:
this method "assigns" a vector to centroids using an rng rule right? If so, can we name it a bit more properly?